File: modules/paycode_event_sets.py
import streamlit as st
import pandas as pd
import requests
import io

from modules.ui_helpers import module_header, section_header


# ======================================================
# PAYCODE EVENT SETS UI
# ======================================================
def paycode_event_sets_ui():

    module_header(
        "🧩 Paycode Event Sets",
        "Create, update, delete and download Paycode Event Sets"
    )

    # ==================================================
    # PRECHECK
    # ==================================================
    if not st.session_state.get("token"):
        st.error("Please login first")
        return

    HOST = st.session_state.HOST.rstrip("/")

    SETS_URL = f"{HOST}/resource-server/api/paycode_event_sets"
    FULL_SETS_URL = f"{HOST}/resource-server/api/paycode_event_sets/?projection=FULL"
    EVENTS_URL = f"{HOST}/resource-server/api/paycode_events"

    headers = {
        "Authorization": f"Bearer {st.session_state.token}",
        "Content-Type": "application/json",
        "Accept": "application/json"
    }

    # ==================================================
    # 1️⃣ DOWNLOAD TEMPLATE
    # ==================================================
    section_header("📥 Download Upload Template")

    # --------------------------------------------------
    # TEMPLATE FORMAT
    # --------------------------------------------------
    template_df = pd.DataFrame(columns=[
        "id",
        "name",
        "description",
        "PaycodeEvent",
        "Priority"
    ])

    # --------------------------------------------------
    # PAYCODE EVENTS EXPORT
    # --------------------------------------------------
    r1 = requests.get(EVENTS_URL, headers=headers)

    if r1.status_code == 200:

        events_df = pd.DataFrame([
            {
                "id": e.get("id"),
                "name": e.get("name"),
                "description": e.get("description")
            }
            for e in r1.json()
        ])

    else:

        events_df = pd.DataFrame(columns=[
            "id",
            "name",
            "description"
        ])

    # --------------------------------------------------
    # EXCEL OUTPUT
    # --------------------------------------------------
    output = io.BytesIO()

    with pd.ExcelWriter(output, engine="openpyxl") as writer:

        # Upload Template
        template_df.to_excel(
            writer,
            index=False,
            sheet_name="Upload_Template"
        )

        # Paycode Event Export
        events_df.to_excel(
            writer,
            index=False,
            sheet_name="Paycode_Events"
        )

    st.download_button(
        "⬇️ Download Template",
        data=output.getvalue(),
        file_name="paycode_event_sets_template.xlsx",
        use_container_width=True
    )

    st.divider()

    # ==================================================
    # 2️⃣ UPLOAD
    # ==================================================
    section_header("📤 Upload Paycode Event Sets")

    uploaded_file = st.file_uploader(
        "Upload CSV or Excel",
        ["csv", "xlsx", "xls"]
    )

    if uploaded_file:

        # --------------------------------------------------
        # READ FILE
        # --------------------------------------------------
        if uploaded_file.name.endswith(".csv"):
            df = pd.read_csv(uploaded_file)
        else:
            df = pd.read_excel(uploaded_file)

        df = df.fillna("")

        st.success(f"Rows detected: {len(df)}")

        st.dataframe(df, use_container_width=True)

        # ==================================================
        # PROCESS
        # ==================================================
        if st.button(
            "🚀 Process Upload",
            type="primary",
            use_container_width=True
        ):

            results = []

            # --------------------------------------------------
            # HELPERS
            # --------------------------------------------------
            def parse_int(x):
                try:
                    return int(float(x))
                except:
                    return None

            # --------------------------------------------------
            # NORMALIZE
            # --------------------------------------------------
            df["id"] = df["id"].apply(parse_int)

            # --------------------------------------------------
            # UPDATE GROUPS
            # --------------------------------------------------
            update_groups = df[
                df["id"].notna()
            ].groupby("id")

            # --------------------------------------------------
            # CREATE GROUPS
            # --------------------------------------------------
            create_groups = df[
                df["id"].isna()
            ].groupby("name")

            with st.spinner("⏳ Processing..."):

                # ==================================================
                # UPDATE EXISTING
                # ==================================================
                for set_id, group in update_groups:

                    try:

                        first = group.iloc[0]

                        payload = {
                            "id": int(set_id),
                            "name": str(first["name"]).strip(),
                            "description": (
                                str(first["description"]).strip()
                                or str(first["name"]).strip()
                            ),
                            "entries": []
                        }

                        seen = set()

                        for _, row in group.iterrows():

                            paycode_event_id = parse_int(
                                row.get("PaycodeEvent")
                            )

                            if not paycode_event_id:
                                continue

                            if paycode_event_id in seen:
                                continue

                            seen.add(paycode_event_id)

                            payload["entries"].append({
                                "paycodeEvent": {
                                    "id": paycode_event_id
                                },
                                "priority": (
                                    parse_int(row.get("Priority")) or 1
                                ),
                                "overridable": False
                            })

                        r = requests.put(
                            f"{SETS_URL}/{int(set_id)}",
                            headers=headers,
                            json=payload
                        )

                        results.append({
                            "id": set_id,
                            "name": payload["name"],
                            "action": "UPDATE",
                            "entries": len(payload["entries"]),
                            "status": (
                                "Success"
                                if r.status_code in (200, 201)
                                else f"Failed ({r.status_code})"
                            )
                        })

                    except Exception as e:

                        results.append({
                            "id": set_id,
                            "name": "",
                            "action": "UPDATE",
                            "entries": "",
                            "status": str(e)
                        })

                # ==================================================
                # CREATE NEW
                # ==================================================
                for name, group in create_groups:

                    try:

                        first = group.iloc[0]

                        payload = {
                            "name": str(name).strip(),
                            "description": (
                                str(first["description"]).strip()
                                or str(name).strip()
                            ),
                            "entries": []
                        }

                        seen = set()

                        for _, row in group.iterrows():

                            paycode_event_id = parse_int(
                                row.get("PaycodeEvent")
                            )

                            if not paycode_event_id:
                                continue

                            if paycode_event_id in seen:
                                continue

                            seen.add(paycode_event_id)

                            payload["entries"].append({
                                "paycodeEvent": {
                                    "id": paycode_event_id
                                },
                                "priority": (
                                    parse_int(row.get("Priority")) or 1
                                ),
                                "overridable": False
                            })

                        r = requests.post(
                            SETS_URL,
                            headers=headers,
                            json=payload
                        )

                        results.append({
                            "id": "",
                            "name": payload["name"],
                            "action": "CREATE",
                            "entries": len(payload["entries"]),
                            "status": (
                                "Success"
                                if r.status_code in (200, 201)
                                else f"Failed ({r.status_code})"
                            )
                        })

                    except Exception as e:

                        results.append({
                            "id": "",
                            "name": name,
                            "action": "CREATE",
                            "entries": "",
                            "status": str(e)
                        })

            # ==================================================
            # RESULT
            # ==================================================
            section_header("📊 Upload Result")

            st.dataframe(
                pd.DataFrame(results),
                use_container_width=True
            )

    st.divider()

    # ==================================================
    # 3️⃣ DELETE
    # ==================================================
    section_header("🗑️ Delete Paycode Event Sets")

    ids_input = st.text_input(
        "Enter IDs (comma-separated)",
        placeholder="143,144"
    )

    if st.button(
        "Delete Paycode Event Sets",
        use_container_width=True
    ):

        ids = [
            i.strip()
            for i in ids_input.split(",")
            if i.strip().isdigit()
        ]

        with st.spinner("⏳ Deleting..."):

            for sid in ids:

                r = requests.delete(
                    f"{SETS_URL}/{sid}",
                    headers=headers
                )

                if r.status_code in (200, 204):
                    st.success(f"Deleted ID {sid}")
                else:
                    st.error(f"Failed to delete ID {sid}")

    st.divider()

    # ==================================================
    # 4️⃣ EXPORT EXISTING
    # ==================================================
    section_header("⬇️ Download Existing Paycode Event Sets")

    with st.spinner("⏳ Fetching..."):

        r = requests.get(
            FULL_SETS_URL,
            headers=headers
        )

        if r.status_code != 200:
            st.error("Failed to fetch data")
            return

        rows = []

        for item in r.json():

            entries = item.get("entries", [])

            for entry in entries:

                rows.append({

                    "id": item.get("id"),

                    "name": item.get("name"),

                    "description": item.get("description"),

                    "PaycodeEvent": (
                        entry.get("paycodeEvent", {}).get("id")
                    ),

                    "Priority": (
                        entry.get("priority")
                    )

                })

        export_df = pd.DataFrame(rows)

        output = io.BytesIO()

        with pd.ExcelWriter(output, engine="openpyxl") as writer:

            export_df.to_excel(
                writer,
                index=False,
                sheet_name="Paycode_Event_Sets"
            )

    st.download_button(
        "⬇️ Download Existing Paycode Event Sets",
        data=output.getvalue(),
        file_name="paycode_event_sets_export.xlsx",
        use_container_width=True
    )

File: renderer/paycode-event-sets.js
/* =========================================================
 * paycode-event-sets.js — BeeForce Configuration Portal
 * Plain script body. Executed as:
 *   new Function('AppCtx', <this file> + '\nreturn (typeof init==="function")?init(AppCtx):undefined;')
 * Defines a top-level `function init(ctx)` — nothing else is required.
 * No import/export syntax. ctx === AppCtx.
 * ========================================================= */
// =========================================================
// modules/paycode-event-sets.js — BeeForce Configuration Portal
// Owns: screen-pes, screen-pes-create, screen-pes-view, screen-pes-delete
// GET  /api/attendance/paycode_event_sets/?projection=FULL
// POST /api/attendance/paycode_event_sets/          (create)
// PUT  /api/attendance/paycode_event_sets/{id}      (update)
// DELETE /api/attendance/paycode_event_sets/{id}
// Exports: init(context)
// =========================================================

function init(ctx) {
  const { AppState, apiH, downloadExcel, wireFileInput, parseApiError, parseIntSafe, Notifications, Router, Audit, ApiTracker, fetchPaycodeEventsRef, modal } = ctx;
  const BASE = () => `${AppState.BASE_URL}/api/attendance/paycode_event_sets`;

  if (!document.getElementById("screen-pes")) {
    const main = document.createElement("div");
    main.id = "screen-pes"; main.style.display = "none";
    main.innerHTML = `
      <div class="bf-breadcrumb"><span class="bc-link" data-go="screen-hub">🏠 Home</span><span class="bc-sep">›</span><span class="bc-cur">🗂️ Event Sets</span></div>
      <div class="bf-module-header"><div class="mh-icon">🗂️</div><div><div class="mh-title">Paycode Event Sets</div><div class="mh-sub">Create · Update · View · Delete</div></div></div>
      <div class="bf-action-grid">
        <div class="bf-action-card" id="pes-action-tpl"><div class="ac-icon">⬇️</div><div class="ac-name">Download Template</div></div>
        <div class="bf-action-card" id="pes-action-create"><div class="ac-icon">📤</div><div class="ac-name">Upload & Create/Update</div></div>
        <div class="bf-action-card" id="pes-action-view"><div class="ac-icon">📋</div><div class="ac-name">View Existing</div></div>
        <div class="bf-action-card" id="pes-action-delete"><div class="ac-icon">🗑️</div><div class="ac-name">Delete</div></div>
      </div>`;
    modal.appendChild(main);

    const create = document.createElement("div");
    create.id = "screen-pes-create"; create.style.display = "none";
    create.innerHTML = `
      <div class="bf-breadcrumb"><span class="bc-link" data-go="screen-hub">🏠 Home</span><span class="bc-sep">›</span><span class="bc-link" data-go="screen-pes">🗂️ Event Sets</span><span class="bc-sep">›</span><span class="bc-cur">Create/Update</span></div>
      <div class="bf-info-box">Leave <b>set_id</b> blank to create. Fill to update. Multiple rows = multiple entries.</div>
      <div id="pes-create-status" class="bf-status"></div>
      <label class="bf-file-label" for="pes-file"><span id="pes-file-name">📂 Click to select filled template (.xlsx)</span></label>
      <input type="file" id="pes-file" class="bf-file-input" accept=".xlsx,.xls,.csv" />
      <div class="bf-progress" id="pes-c-prog"><div class="bf-pb" id="pes-c-bar" style="width:0%"></div></div>
      <div class="bf-log" id="pes-create-log"></div>
      <button class="bf-btn bf-success" id="pes-submit" disabled>🚀 Submit Paycode Event Sets</button>`;
    modal.appendChild(create);

    const view = document.createElement("div");
    view.id = "screen-pes-view"; view.style.display = "none";
    view.innerHTML = `
      <div class="bf-breadcrumb"><span class="bc-link" data-go="screen-hub">🏠 Home</span><span class="bc-sep">›</span><span class="bc-link" data-go="screen-pes">🗂️ Event Sets</span><span class="bc-sep">›</span><span class="bc-cur">View</span></div>
      <div id="pes-view-status" class="bf-status"></div>
      <div class="bf-progress" id="pes-v-prog" style="display:none"><div class="bf-pb" id="pes-v-bar" style="width:0%"></div></div>
      <button class="bf-btn bf-primary" id="pes-fetch">⬇️ Download Event Sets</button>`;
    modal.appendChild(view);

    const del = document.createElement("div");
    del.id = "screen-pes-delete"; del.style.display = "none";
    del.innerHTML = `
      <div class="bf-breadcrumb"><span class="bc-link" data-go="screen-hub">🏠 Home</span><span class="bc-sep">›</span><span class="bc-link" data-go="screen-pes">🗂️ Event Sets</span><span class="bc-sep">›</span><span class="bc-cur">Delete</span></div>
      <div class="bf-warn-box">⚠️ Deletion is permanent.</div>
      <div id="pes-delete-status" class="bf-status"></div>
      <label class="bf-label">Set IDs to delete (comma-separated)</label>
      <input type="text" id="pes-del-ids" placeholder="143, 144" />
      <label class="bf-label" style="display:flex;align-items:center;gap:8px;cursor:pointer;font-weight:400;"><input type="checkbox" id="pes-confirm" style="width:auto;margin:0;" />I understand this is permanent</label>
      <div class="bf-log" id="pes-delete-log"></div>
      <button class="bf-btn bf-danger" id="pes-do-delete" style="margin-top:10px;">🗑️ Delete</button>`;
    modal.appendChild(del);

    Router.register("screen-pes", "screen-pes-create", "screen-pes-view", "screen-pes-delete");
    [main, create, view, del].forEach(el => {
      Router.wireDataGoLinks(el);
      el.querySelectorAll("input,textarea").forEach(inp =>
        ["keydown", "keyup", "keypress", "input"].forEach(evt => inp.addEventListener(evt, e => e.stopPropagation(), true))
      );
    });
  }

  document.getElementById("pes-action-tpl").onclick = async () => {
    const btn = document.getElementById("pes-action-tpl"); btn.style.opacity = "0.6";
    const events = await fetchPaycodeEventsRef();
    downloadExcel("paycode_event_sets_template.xlsx", [
      { name: "Event_Sets", tabColor: "7C3AED", headers: ["set_id", "set_name", "set_description", "entry_id", "priority", "paycode_event_id"], rows: [], highlightCols: [1, 5] },
      { name: "Paycode_Events_Master", tabColor: "059669", headers: ["id", "name", "description", "paycode_id"], rows: events.map(e => [e.id, e.name || "", e.description || "", e.paycode?.id || ""]) }
    ]);
    btn.style.opacity = "1";
  };
  document.getElementById("pes-action-create").onclick = () => {
    Notifications.clearLog("pes-create-log"); AppState.cache.pes = null;
    document.getElementById("pes-submit").disabled = true;
    document.getElementById("pes-file-name").textContent = "📂 Click to select filled template (.xlsx)";
    Router.go("screen-pes-create");
  };
  document.getElementById("pes-action-view").onclick = () => Router.go("screen-pes-view");
  document.getElementById("pes-action-delete").onclick = () => { Notifications.clearLog("pes-delete-log"); Router.go("screen-pes-delete"); };
  wireFileInput("pes-file", "pes-file-name", "pes-submit", "pes");

  document.getElementById("pes-submit").onclick = async () => {
    const btn = document.getElementById("pes-submit");
    const rows = AppState.cache.pes || [];
    Notifications.clearLog("pes-create-log");
    if (!rows.length) { Notifications.error("pes-create-status", "❌ No data found"); return; }

    const updateGroups = {}, createGroups = {};
    rows.forEach(row => {
      const setId = parseIntSafe(row["set_id"]);
      if (setId !== null) { if (!updateGroups[setId]) updateGroups[setId] = []; updateGroups[setId].push(row); }
      else { const setName = String(row["set_name"] || "").trim(); if (!setName) return; if (!createGroups[setName]) createGroups[setName] = []; createGroups[setName].push(row); }
    });
    const updateEntries = Object.entries(updateGroups), createEntries = Object.entries(createGroups);
    const total = updateEntries.length + createEntries.length;
    if (!total) { Notifications.error("pes-create-status", "❌ No valid groups found"); return; }

    Notifications.loading(btn, "Submitting...");
    Notifications.progress.show("pes-c-prog");
    let s = 0, f = 0, done = 0;

    for (const [setId, group] of updateEntries) {
      const first = group[0];
      const payload = { id: parseInt(setId), name: String(first["set_name"] || "").trim(), description: String(first["set_description"] || "").trim() || String(first["set_name"] || "").trim(), entries: [] };
      const seen = new Set();
      group.forEach(row => {
        const entryId = parseIntSafe(row["entry_id"]); const priority = parseIntSafe(row["priority"]) || 1; const pcEventId = parseIntSafe(row["paycode_event_id"]);
        if (!pcEventId) return; const key = `${entryId}-${pcEventId}`; if (seen.has(key)) return; seen.add(key);
        const entry = { paycodeEvent: { id: pcEventId }, priority, overridable: false }; if (entryId !== null) entry.id = entryId;
        payload.entries.push(entry);
      });
      try {
        const r = await ApiTracker.fetch(`${BASE()}/${payload.id}`, { method: "PUT", headers: apiH(), body: JSON.stringify(payload) }, "Paycode Event Sets");
        if (r.ok) { Notifications.log("pes-create-log", `✅ Updated: ${payload.name} (${payload.entries.length} entries)`, "ok"); s++; Audit.onDbOp("updated", 1); }
        else { Notifications.log("pes-create-log", `❌ Failed update: ${payload.name} — ${parseApiError(await r.text())}`, "fail"); f++; }
      } catch (e) { Notifications.log("pes-create-log", `❌ Error: ${payload.name} — ${e.message}`, "fail"); f++; }
      done++; Notifications.progress.set("pes-c-bar", (done / total) * 100);
    }
    for (const [setName, group] of createEntries) {
      const first = group[0];
      const payload = { name: setName, description: String(first["set_description"] || "").trim() || setName, entries: [] };
      const seen = new Set();
      group.forEach(row => {
        const priority = parseIntSafe(row["priority"]) || 1; const pcEventId = parseIntSafe(row["paycode_event_id"]);
        if (!pcEventId || seen.has(pcEventId)) return; seen.add(pcEventId);
        payload.entries.push({ paycodeEvent: { id: pcEventId }, priority, overridable: false });
      });
      try {
        const r = await ApiTracker.fetch(`${BASE()}/`, { method: "POST", headers: apiH(), body: JSON.stringify(payload) }, "Paycode Event Sets");
        if (r.ok) { Notifications.log("pes-create-log", `✅ Created: ${payload.name} (${payload.entries.length} entries)`, "ok"); s++; Audit.onDbOp("created", 1); }
        else { Notifications.log("pes-create-log", `❌ Failed create: ${payload.name} — ${parseApiError(await r.text())}`, "fail"); f++; }
      } catch (e) { Notifications.log("pes-create-log", `❌ Error: ${payload.name} — ${e.message}`, "fail"); f++; }
      done++; Notifications.progress.set("pes-c-bar", (done / total) * 100);
    }

    Notifications.log("pes-create-log", `━━━━━━━━━━━━━━━━━━`, "info");
    Notifications.log("pes-create-log", `✅ Success: ${s}  ❌ Failed: ${f}`, f === 0 ? "ok" : "info");
    Notifications.status("pes-create-status", `Done — ${s} success, ${f} failed`, f === 0 ? "success" : "info");
    Notifications.doneLoading(btn, "🚀 Submit Paycode Event Sets");
    Notifications.progress.hide("pes-c-prog");
  };

  document.getElementById("pes-do-delete").onclick = async () => {
    if (!document.getElementById("pes-confirm").checked) { Notifications.error("pes-delete-status", "⚠️ Check confirmation box"); return; }
    const ids = document.getElementById("pes-del-ids").value.split(",").map(s => s.trim()).filter(s => !isNaN(s) && s !== "");
    if (!ids.length) { Notifications.error("pes-delete-status", "❌ No valid IDs"); return; }
    const btn = document.getElementById("pes-do-delete");
    Notifications.loading(btn, "Deleting..."); Notifications.clearLog("pes-delete-log");
    let s = 0, f = 0;
    for (const id of ids) {
      try {
        const r = await ApiTracker.fetch(`${BASE()}/${id}`, { method: "DELETE", headers: apiH() }, "Paycode Event Sets");
        if (r.status === 200 || r.status === 204) { Notifications.log("pes-delete-log", `✅ Deleted: ${id}`, "ok"); s++; }
        else { Notifications.log("pes-delete-log", `❌ Failed: ${id}`, "fail"); f++; }
      } catch (e) { Notifications.log("pes-delete-log", `❌ Error: ${id}`, "fail"); f++; }
    }
    Audit.onDbOp("deleted", s);
    Notifications.status("pes-delete-status", `Done — ${s} deleted, ${f} failed`, f === 0 ? "success" : "info");
    Notifications.doneLoading(btn, "🗑️ Delete");
  };

  document.getElementById("pes-fetch").onclick = async () => {
    const btn = document.getElementById("pes-fetch");
    Notifications.loading(btn, "Connecting to server...");
    Notifications.progress.show("pes-v-prog");
    try {
      Notifications.info("pes-view-status", "⏳ Fetching records..."); Notifications.progress.set("pes-v-bar", 25);
      const r = await ApiTracker.fetch(`${BASE()}/?projection=FULL`, { headers: apiH() }, "Paycode Event Sets");
      if (!r.ok) throw new Error(`HTTP ${r.status}: ${parseApiError(await r.text())}`);
      AppState.cache.pesView = await r.json();
      Notifications.info("pes-view-status", `⏳ Processing ${AppState.cache.pesView.length} set(s)...`); Notifications.progress.set("pes-v-bar", 60);
      const rows = [];
      (AppState.cache.pesView || []).forEach(set => (set.entries || []).forEach(entry => {
        const pe = entry.paycodeEvent || {};
        rows.push([set.id, set.name || "", set.description || "", pe.id || "", entry.priority || "", entry.id || "", pe.name || "", pe.description || "", pe.paycode?.id || "", pe.paycode?.code || ""]);
      }));
      if (!rows.length) { Notifications.info("pes-view-status", "ℹ️ No records available to download."); return; }
      Notifications.info("pes-view-status", "⏳ Preparing Excel..."); Notifications.progress.set("pes-v-bar", 85);
      downloadExcel("paycode_event_sets_existing.xlsx", [{
        name: "Event_Sets", tabColor: "7C3AED",
        headers: ["id", "name", "description", "paycode_event_id", "priority", "entry_id", "paycode_event_name", "paycode_event_description", "paycode_id", "paycode_code"],
        rows
      }]);
      Notifications.progress.set("pes-v-bar", 100);
      Notifications.success("pes-view-status", `✅ Download complete — ${AppState.cache.pesView.length} set(s)`);
    } catch (e) {
      Notifications.error("pes-view-status", `❌ ${e.message}`); Notifications.progress.set("pes-v-bar", 0);
    } finally {
      Notifications.doneLoading(btn, "⬇️ Download Event Sets");
      Notifications.progress.hide("pes-v-prog", 2000);
    }
  };
}
